protocol/headers.go

```diff
@@ -12,162 +12,464 @@
+// If there are more than one headers differing only in capitalization, the ContentType returns the first met value.
+// To use the provided key to get the value, access the map directly.
+func (h Headers) ContentType() string {
+	if value, ok := h[HeaderContentType]; ok {
+		str, _ := h.stringValue(value, "")
+		return str
+	}
+	keys := sortHeadersKey(h)
+	for _, k := range keys {
+		if strings.EqualFold(k, HeaderContentType) {
+			str, _ := h.stringValue(h[k], "")
+			return str
+		}
 	}
-	return h.Values[HeaderIfNoneMatch].(string)
+	return ""
 }
 
-// ReplyTarget returns the 'ditto-reply-target' header value or empty string if not set.
-func (h *Headers) ReplyTarget() int64 {
-	if h.Values[HeaderReplyTarget] == nil {
-		return 0
+// Generic returns the value of the provided key header.
+func (h Headers) Generic(id string) interface{} {
+	return h[id]
+}
+
+// With sets new Headers to the existing.
+func (h Headers) With(opts ...HeaderOpt) Headers {
+	res := make(map[string]interface{})
+
+	for key, value := range h {
+		res[key] = value
 	}
-	return h.Values[HeaderReplyTarget].(int64)
+
+	if err := applyOptsHeader(res, opts...); err != nil {
+		return nil
+	}
+	return res
 }
 
-// ReplyTo returns the 'reply-to' header value or empty string if not set.
-func (h *Headers) ReplyTo() string {
-	if h.Values[HeaderReplyTo] == nil {
-		return ""
+func (h Headers) stringValue(headerValue interface{}, defValue string) (string, bool) {
+	if value, ok := headerValue.(string); ok {
+		return value, true
 	}
-	return h.Values[HeaderReplyTo].(string)
+	return defValue, false
 }
 
-// Version returns the 'version' header value or empty string if not set.
-func (h *Headers) Version() int64 {
-	if h.Values[HeaderSchemaVersion] == nil {
-		return 0
+func (h Headers) timeoutValue(headerValue interface{}) time.Duration {
+	if value, ok := headerValue.(string); ok {
+		if duration, err := parseTimeout(value); err == nil {
+			return duration
+		}
 	}
-	return h.Values[HeaderSchemaVersion].(int64)
+	return 60 * time.Second
 }
 
-// ContentType returns the 'content-type' header value or empty string if not set.
-func (h *Headers) ContentType() string {
-	if h.Values[HeaderContentType] == nil {
-		return ""
+func (h Headers) intValue(headerValue interface{}, defValue int64) int64 {
+	if value, ok := headerValue.(int64); ok {
+		return value
 	}
-	return h.Values[HeaderContentType].(string)
+	return defValue
 }
 
-// Generic returns the value of the provided key header and if a header with such key is present.
-func (h *Headers) Generic(id string) interface{} {
-	return h.Values[id]
+func (h Headers) boolValue(headerValue interface{}, defValue bool) bool {
+	if value, ok := headerValue.(bool); ok {
+		return value
+	}
+	return defValue
 }
 
-// MarshalJSON marshels Headers.
-func (h *Headers) MarshalJSON() ([]byte, error) {
-	return json.Marshal(h.Values)
+func sortHeadersKey(h Headers) []string {
+	var keys []string
+	for k := range h {
+		keys = append(keys, k)
```

Comment:
It would be more efficient to allocate the []string in advance as the size is known, than using append.